File: leetcode/Hand of Straights.js
var isNStraightHand = function (hand, groupSize) {
  const n = hand.length
  if (n % groupSize) return false

  const counter = new Map()
  for (let item of hand) {
    const val = counter.get(item) ? counter.get(item) : 0
    counter.set(item, val + 1)
  }

  const keys = [...counter.keys()].sort((a, b) => a - b)
  let minKeyIdx = 0
  const numGroup = Math.floor(n / groupSize)

  for (let i = 0; i < numGroup; i++) {
    while (minKeyIdx < keys.length && !counter.get(keys[minKeyIdx])) {
      minKeyIdx++
    }
    if (minKeyIdx >= keys.length) {
      return false
    }
    let key = keys[minKeyIdx]
    for (let i = 0; i < groupSize; i++) {
      if (!counter.get(key + i)) {
        return false
      }
      counter.set(key + i, counter.get(key + i) - 1)
    }
  }

  return true
}

console.log(isNStraightHand([8, 8, 9, 7, 7, 7, 6, 7, 10, 6], 2) === true)
console.log(isNStraightHand([1, 2, 3, 6, 2, 3, 4, 7, 8], 3) === true)
console.log(isNStraightHand([8, 10, 12], 3) === false)
console.log(isNStraightHand([1, 2, 3, 4, 5], 4) === false)

File: leetcode/Hand of Straights.py
# https://leetcode.com/problems/hand-of-straights/submissions/
from collections import Counter
from typing import List

from tester import Tester


class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        n = len(hand)
        if n % groupSize:
            return False

        counter = Counter(hand)
        keys = sorted(counter.keys())
        minKeyIdx = 0
        numGroup = n // groupSize

        for i in range(numGroup):
            while minKeyIdx < len(keys) and not counter[keys[minKeyIdx]]:
                minKeyIdx += 1
            if minKeyIdx >= len(keys):
                return False

            key = keys[minKeyIdx]
            for i in range(groupSize):
                if not counter[key + i]:
                    return False
                counter[key + i] = counter[key + i] - 1
        return True


t = Tester(Solution())

t.test(True, [1, 2, 3], 3)
t.test(True, [8, 8, 9, 7, 7, 7, 6, 7, 10, 6], 2)
t.test(True, [1, 2, 3, 6, 2, 3, 4, 7, 8], 3)
t.test(False, [8, 10, 12], 3)
t.test(False, [1, 2, 3, 4, 5], 4)

t.report()
